modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -4350,15 +4940,357 @@ void GDScriptAnalyzer::reduce_unary_op(GDScriptParser::UnaryOpNode *p_unary_op)
+	return false;
+}
+
+Vector<GDScriptParser::IdentifierNode *> GDScriptAnalyzer::deduce_nullable_narrowing(GDScriptParser::ExpressionNode *p_node, bool p_should_be_false) {
+	if (p_node == nullptr) {
+		return {};
+	}
+
+	switch (p_node->type) {
+		case GDScriptParser::Node::BINARY_OPERATOR: {
+			// This deduces that "!=" and "==" can be used to narrow nullables.
+			GDScriptParser::BinaryOpNode *binary_op = static_cast<GDScriptParser::BinaryOpNode *>(p_node);
+			GDScriptParser::ExpressionNode *left = binary_op->left_operand;
+			GDScriptParser::ExpressionNode *right = binary_op->right_operand;
+
+			if (!left || !right) {
+				break;
+			}
+			switch (binary_op->variant_op) {
+				case Variant::OP_NOT_EQUAL:
+				case Variant::OP_EQUAL: {
+					if ((left->type != GDScriptParser::Node::LITERAL && left->type != GDScriptParser::Node::IDENTIFIER) && (right->type != GDScriptParser::Node::LITERAL && right->type != GDScriptParser::Node::IDENTIFIER)) {
+						// At least one of the operands must be a literal or an identifier
+						break;
+					}
+					if (left->get_datatype().is_nullable && right->get_datatype().is_nullable) {
+						// If both operands are nullable, we can't narrow them
+						break;
+					}
+
+					return find_nullable_identifier(left, right, (p_should_be_false ^ (binary_op->variant_op == Variant::OP_EQUAL)));
+				} break;
+				case Variant::OP_AND: {
+					// For the AND operator we can try deducing each side and then combining the resulting identifiers.
+					// That way we can support situations like: "if foo and bar:", where both foo and bar will be narrowed
+					Vector<GDScriptParser::IdentifierNode *> left_identifiers = deduce_nullable_narrowing(left, p_should_be_false);
+					Vector<GDScriptParser::IdentifierNode *> right_identifiers = deduce_nullable_narrowing(right, p_should_be_false);
+					// This might end up repeating identifiers, but given they will ultimately be stored in a HashSet it should be fine
+					left_identifiers.append_array(right_identifiers);
+					return left_identifiers;
+				} break;
+				default: {
+					return {}; // Not supported
+				}
+			}
+
+		} break;
+		case GDScriptParser::Node::IDENTIFIER: {
+			// This deduces that any sole identifier must be narrowed.
+			if (!p_should_be_false) {
+				return { static_cast<GDScriptParser::IdentifierNode *>(p_node) };
+			}
+		} break;
+		case GDScriptParser::Node::UNARY_OPERATOR: {
+			// This deduces that, except for the "!" operator (logic not), unary operators will always narrow nullables.
+			GDScriptParser::UnaryOpNode *unary_op = static_cast<GDScriptParser::UnaryOpNode *>(p_node);
+
+			if (!unary_op->operand) {
+				break;
+			}
+
+			if (!unary_op->operand->get_datatype().is_nullable) {
+				break;
+			}
+
+			if (unary_op->operation != GDScriptParser::UnaryOpNode::OP_LOGIC_NOT) {
+				break;
+			}
+
+			if (p_should_be_false) {
+				return { static_cast<GDScriptParser::IdentifierNode *>(unary_op->operand) };
```

Comment:
This crashes as there's no guarantees that `not x` is always an identifier, it could be `not some_fn()`. This could be fixed by delegating to another call of `return deduce_nullable_narrowing(unary_op->operand)`.